src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala

```diff
@@ -0,0 +1,181 @@
+  }
+
+  def killUnknownTaskById(taskId: Task.Id): Unit = {
+    log.debug("Received KillUnknownTaskById({})", taskId)
+    tasksToKill.update(taskId, None)
+    processKills()
+  }
+
+  def killTasks(tasks: Iterable[Task], promise: Promise[Unit]): Unit = {
+    log.debug("Adding {} tasks to queue; setting up child actor to track progress", tasks.size)
+    context.actorOf(TaskKillProgressActor.props(tasks.map(_.taskId), promise))
+    tasks.foreach { task =>
+      tasksToKill.update(task.taskId, Some(task))
+    }
+    processKills()
+  }
+
+  private def processKills(): Unit = {
+    val killCount = config.killChunkSize - inFlight.size
+    val toKillNow = tasksToKill.take(killCount)
+
+    log.info("processing {} kills", toKillNow.size)
+    toKillNow.foreach {
+      case (taskId, maybeTask) => processKill(taskId, maybeTask)
+    }
+
+    retryTimer = Some(context.system.scheduler.scheduleOnce(config.killRetryTimeout, self, Retry))
+  }
+
+  def processKill(taskId: Task.Id, maybeTask: Option[Task]): Unit = {
+    val taskIsLost: Boolean = maybeTask.fold(false)(isLost)
+
+    if (taskIsLost) {
+      log.warning("Expunging lost {} from state because it should be killed", taskId)
+      // we will eventually be notified of a taskStatusUpdate after the task has been expunged
+      stateOpProcessor.process(TaskStateOp.ForceExpunge(taskId))
+    } else {
+      val knownOrNot = if (maybeTask.isDefined) "known" else "unknown"
+      log.warning("Killing {} {}", knownOrNot, taskId)
+      driverHolder.driver.map(_.killTask(taskId.mesosTaskId))
+    }
+
+    val attempts = inFlight.get(taskId).fold(1)(_.attempts + 1)
+    inFlight.update(taskId, TaskToKill(taskId, maybeTask, issued = clock.now(), attempts))
+    tasksToKill.remove(taskId)
+  }
+
+  def handleTerminal(taskId: Task.Id): Unit = {
+    tasksToKill.remove(taskId)
+    inFlight.remove(taskId)
+    log.debug("{} is terminal. ({} kills queued, {} in flight)", taskId, tasksToKill.size, inFlight.size)
+    processKills()
+  }
+
+  def retry(): Unit = {
+    val now = clock.now()
+
+    inFlight.foreach {
+      case (taskId, taskToKill) if taskToKill.attempts >= config.killRetryMax =>
+        log.warning("Expunging {} from state: max retries reached", taskId)
+        stateOpProcessor.process(TaskStateOp.ForceExpunge(taskId))
+
+      case (taskId, taskToKill) if (taskToKill.issued + config.killRetryTimeout) < now =>
+        log.warning("No kill ack received for {}, retrying", taskId)
+        processKill(taskId, taskToKill.maybeTask)
+
+      case _ => // ignore
+    }
+
+    if (inFlight.nonEmpty) {
+      retryTimer = Some(context.system.scheduler.scheduleOnce(config.killRetryTimeout, self, Retry))
+    } else {
+      retryTimer = None
+    }
+  }
+
+  def isLost(task: Task): Boolean = {
+    import org.apache.mesos
+    task.mesosStatus.fold(false)(_.getState == mesos.Protos.TaskState.TASK_LOST)
+  }
+}
+
+private[termination] object TaskKillServiceActor {
+
+  sealed trait Request
```

Comment:
Sounds good to me

On Sat, Jul 23, 2016 at 5:38 AM, Matthias Eichstedt <
notifications@github.com> wrote:

> In
> src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala
> <https://github.com/mesosphere/marathon/pull/4117#discussion_r71970260>:
>
> > +    if (inFlight.nonEmpty) {
> > +      retryTimer = Some(context.system.scheduler.scheduleOnce(config.killRetryTimeout, self, Retry))
> > +    } else {
> > +      retryTimer = None
> > +    }
> > +  }
> > +
> > +  def isLost(task: Task): Boolean = {
> > +    import org.apache.mesos
> > +    task.mesosStatus.fold(false)(_.getState == mesos.Protos.TaskState.TASK_LOST)
> > +  }
> > +}
> > +
> > +private[termination] object TaskKillServiceActor {
> > +
> > +  sealed trait Request
>
> +1 from me
> I actually thought about that! It always bothers me if the logs don't say
> *why* a task was killed, we have to infer the reason from the logs above
> which can be painful.
>
> Would you be comfortable with a trait TaskKillReason on the delegate
> level of TaskKillService, so that every kill() function has a second
> argument that defines that reason?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/marathon/pull/4117/files/6dfe23f736dca401cdc59579804d1c405ee8b1c0#r71970260>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACPVLAXVHEDS0M0H-WRkOw157wu2NYqMks5qYeEzgaJpZM4JQmUs>
> .
>